pkg/frontend/querymiddleware/labels_query_optimizer.go

```diff
@@ -166,14 +168,21 @@ func optimizeLabelsRequestMatchers(rawMatcherSets []string) (_ []string, optimiz
 			optimizedMatchers = append(optimizedMatchers, matcher)
 		}
 
-		if len(optimizedMatchers) > 0 {
-			optimizedRawMatcherSets = append(optimizedRawMatcherSets, util.LabelMatchersToString(optimizedMatchers))
-		} else {
-			// It was not an empty matcher (see condition above), and all the matchers have been removed.
+		if !hasNonEmptyMatchers {
```

Comment:
if the request has `{foo=~".*", bar=~".*"}` then your implementation would forward these to downstream. Shouldn’t it remove both instead?